impl/src/main/java/slack/ReactionMessageHandler.java

```diff
@@ -18,66 +20,96 @@
     }
 
     @Override
     public Observable<Void> handleMessage(JsonObject message) {
-        final String threadId = message.get("item").getAsJsonObject().get("ts").getAsString();
-
-        final AtomicReference<Boolean> upVote = new AtomicReference<>();
-        String reaction = message.get("reaction").getAsString();
-
-        if(POSITIVE_EMOJIS.contains(reaction)) {
-            upVote.set(true);
-        }
-
-        if(NEGATIVE_EMOJIS.contains(reaction)) {
-            upVote.set(false);
-        }
 
+        final AtomicReference<Boolean> upVote = getUpVoteValue(message);
         //No recognized reaction
-        if(upVote.get() == null) {
+        if (upVote.get() == null) {
             return empty();
         }
 
         //Reaction removed -> invert the vote
-        if(REACTION_REMOVED.equals(message.get("type").getAsString())) {
+        if (REACTION_REMOVED.equals(message.get(TYPE).getAsString())) {
             upVote.set(!upVote.get());
         }
 
-        return questionResource.getQuestionBySlackThreadId(threadId)
-            .onErrorResumeNext(throwable -> {
-                //thread id is not a question but an answer
-                if(NOT_FOUND.equals(((WebException)throwable).getStatus())) {
-                    return upVote.get() ? answerResource.upVoteAnswer(threadId).cast(Question.class) : answerResource.downVoteAnswer(threadId).cast(Question.class);
-                }
-                return error(throwable);
-            })
-            .flatMap(question -> upVote.get() ? questionResource.upVoteQuestion(threadId) : questionResource.downVoteQuestion(threadId));
+        final String threadId = getThread(message);
+        return slackResource.getMessageFromSlack(getChannel(message), getThread(message))
+            .flatMap(mainMessage -> getUserId(mainMessage)
+                .flatMap(userId -> questionResource.getQuestionBySlackThreadId(threadId)
+                    .onErrorResumeNext(throwable -> {
+                        //thread id is not a question but an answer
+                        if (NOT_FOUND.equals(((WebException)throwable).getStatus())) {
+                            return answerResource.getAnswerBySlackId(threadId)
+                                .flatMap(answer -> upVote.get() ?
+                                    userAnswerResource.upVoteAnswer(userId, answer.getId()).cast(Question.class) :
+                                    userAnswerResource.downVoteAnswer(userId, answer.getId()).cast(Question.class));
+                        }
+                        return error(throwable);
+                    })
+                    .flatMap(question -> upVote.get() ? questionResource.upVoteQuestion(threadId) : questionResource.downVoteQuestion(threadId))));
+    }
+
+    private static AtomicReference<Boolean> getUpVoteValue(JsonObject message) {
+        AtomicReference<Boolean> upVote   = new AtomicReference<>();
```

Comment:
i dont think there is a reason to have atomic reference here? The reference is only used inside of the method, so race conditions should not occur? If i understand it correctly :) simply return in the if cases :)

Reply:
I think it's because it's used in the lambda and must be "effectively final". I'll see if I can do something else.